### DISCUSSION TITLE
Do we need to fold_in the rng that are requested by make_rng to make it different?

### DISCUSSION THREAD

**Discussion by wztdream** [google/flax#1715] [Q&A] [answered]
Hi,
Below is the doc of `make_rng`,
> Returns a new RNG key from a given RNG sequence for this Module.

>The new RNG key is split from the previous one. Thus, every call to make_rng returns a new RNG key, while still guaranteeing full reproducibility.

 I suppose `make_rng` will do the `fold_in` staff underneath, but in sst2 example, they use `fold_in` to make sure use new rng each step, is it necessary here?
https://github.com/google/flax/blob/4cbf6bfa472149fb2e43c1444154bf113aae5041/examples/sst2/train.py#L097-L112

Here is the model that use this rng, it is variation of dropout:

https://github.com/google/flax/blob/4cbf6bfa472149fb2e43c1444154bf113aae5041/examples/sst2/models.py#L81-L99

**Comment by avital** [comment]
The comment about `make_rng` splitting the PRNG key is true across multiple invocations within a Flax module. Flax doesn't do anything special for RNGs that are passed in directly to JAX jitted functions, such as `train_step` in the example you linked to. There, you're doing "vanilla JAX" and need to manually split. You still need to pass in a unique RNG to each module when you call `apply`, or you will get the same dropout mask. (This is a feature -- we /can/ be fully reproducible)

**Reply by wztdream** [reply]
Thank you

**Comment by marcvanzee** [comment, marked as answer]
In the documentation, the phrase "every call to make_rng returns a new RNG key" means that within a Module, each call to `make_rng` will create a fresh RNG key even if the name is the same, ie.:

```python
from flax import linen as nn
from jax.random import PRNGKey
from jax.numpy import zeros

class Foo(nn.Module):
  @nn.compact
  def __call__(self, x):
    rng1 = self.make_rng('dropout')
    rng2 = self.make_rng('dropout')
    assert (rng1 != rng2).any()

rngs = {'params': PRNGKey(0), 'dropout': PRNGKey(1)}
_ = Foo().init(rngs, zeros((2,)))
```

However, in the SST2 example (and in other examples like [WMT](https://github.com/google/flax/blob/main/examples/wmt/train.py#L189)), we want a fresh RNG *for each time we call the Module*. If we wouldn't do this, then we would be calling the Module with the same rng repeatedly, and given that the RNG splitting is fully deterministic, this would result in re-using the same RNG, which is discouraged by JAX:

```python
class Bar(nn.Module):
  @nn.compact
  def __call__(self, x):
    rng1 = self.make_rng('dropout')
    return rng1

rngs = {'params': PRNGKey(0), 'dropout': PRNGKey(1)}

y1, _ = Bar().init_with_output(rngs, zeros((2,)))
y2, _ = Bar().init_with_output(rngs, zeros((2,)))

assert (y1 == y2).all()
```

**Reply by wztdream** [reply]
Thank you both, this is clearer.
